autoload/echodoc/util.vim

```diff
@@ -0,0 +1,148 @@
+" Returns a parsed stack of functions found in the text.  Each item in the
+" stack contains a dict:
+" - name: Function name.
+" - start: Argument start position.
+" - end: Argument end position.  -1 if the function is unclosed.
+" - pos: The argument position.  1-indexed, 0 = no args, -1 = closed.
+" - ppos: The function's position in the previous function in the stack.
+" - args: A list of argument positions.
+function! echodoc#util#parse_funcs(text) abort
+  if a:text == ''
+    return []
+  endif
+
+  let quote_i = -1
+  let stack = []
+  let open_stack = []
+  let gap = 0
+
+  " Matching pairs will count as a single argument entry so that commas can be
+  " skipped within them.  The open depth is tracked in each open stack item.
+  " Parenthesis is an exception since it's used for functions and can have a
+  " depth of 1.
+  let pairs = '({[)}]'
+  let l = len(a:text) - 1
+  let i = -1
+
+  while i < l
+    let i += 1
+    let c = a:text[i]
+
+    if i > 0 && a:text[i - 1] == '\'
+      continue
+    endif
+
+    if quote_i != -1
+      " For languages that allow '''' ?
+      " if c == "'" && a:text[i - 1] == c && i - quote_i > 1
+      "   continue
+      " endif
+      if c == a:text[quote_i]
+        let quote_i = -1
+      endif
+      continue
+    endif
+
+    if quote_i == -1 && (c == "'" || c == '"' || c == '`')
+      " backtick (`) is not used alone in languages that I know of.
+      let quote_i = i
+      continue
+    endif
+
+    let prev = len(open_stack) ? open_stack[-1] : {'opens': [0, 0, 0]}
+    execute 'let opened = '.join(prev.opens, '+')
```

Comment:
Because I didn't want to type:

```vim
let opened = prev.opens[0] + prev.opens[1] + prev.opens[2]
```

It's changed now because the number of items shouldn't change in the future.